### PR TITLE
GCP Storage Sink Connector

### PR DESCRIPTION
**Pull request by davidsloan**
## Connector Overview

The Kafka sink connector for GCP Storage is a powerful tool that allows you to seamlessly stream data from Apache Kafka topics into your GCP Storage bucket. This connector facilitates real-time data integration and analytics, making it a crucial component of data pipelines. It supports various data formats, compression options, and error handling policies to fit your specific use case.

## Key Features

- **Real-time Data Streaming**: The connector enables real-time data streaming from Kafka topics to GCP Storage, ensuring that your data is readily available for analysis and processing.

- **Data Format Support**: You can choose the data format that best suits your needs, including Avro and Parquet, and configure compression settings.

- **Error Handling**: The connector offers configurable error handling policies, allowing you to determine how errors are managed during data transfer. You can choose to ignore errors (NOOP), propagate them (THROW), or retry (RETRY) with defined retry limits.

- **Flexible Configuration**: The connector provides various configuration options to adapt to your specific GCP Storage setup and requirements. This includes authentication modes, connection settings, and more.

- **File Handling**: You can configure local temporary directories, padding, and flush behavior to optimize the storage process.

## Connector Configuration

Here's an overview of the connector's configuration options:

| Configuration Option                   | Description                                                             | Default Value |
|---------------------------------------|-------------------------------------------------------------------------|---------------|
| `connect.gcpstorage.kcql`               | Specifies the mapping between Kafka topics and GCP Storage using Kafka Connect Query Language. | - |
| `connect.gcpstorage.gcp.project.id`               | GCP Project ID. | - |
| `connect.gcpstorage.gcp.quota.project.id`               | GCP Quota Project ID. | - |
| `connect.gcpstorage.gcp.auth.mode`     | Authentication mode, options are 'credentials', 'file', or 'default'. | Default |
| `connect.gcpstorage.gcp.credentials`   | For "auth.mode" credentials: GCP Authentication credentials string.                                | [hidden]      |
| `connect.gcpstorage.gcp.file`  | For "auth.mode" file: Local file path for file containing GCP authentication credentials.                           | ""            |
| `connect.gcpstorage.compression.codec`   | Compression codec for data formats.                                    | UNCOMPRESSED  |
| `connect.gcpstorage.compression.level`   | Compression level for certain codecs.                                  | -1            |
| `connect.gcpstorage.disable.flush.count` | Option to disable flushing based on record count.                     | false         |
| `connect.gcpstorage.endpoint`            | Endpoint for GCP Storage.                                          | ""            |
| `connect.gcpstorage.http.connection.timeout` | Connection timeout for HTTP requests.                                | 60000         |
| `connect.gcpstorage.http.socket.timeout`  | Socket timeout for HTTP requests.                                      | 60000         |
| `connect.gcpstorage.local.tmp.directory`  | Local temporary directory for file preparation.                        | ""            |
| `connect.gcpstorage.padding.length`       | Padding length for string values.                                       | -1            |
| `connect.gcpstorage.padding.strategy`     | Padding strategy for partition and offset in output files. Options are 'LeftPad', 'RightPad', or 'NoOp' (no padding). Defaults to 'LeftPad'. | "" |
| `connect.gcpstorage.pool.max.connections` | Maximum connections in the pool. -1: Use default according to the underlying client. | -1 |
| `connect.gcpstorage.error.policy`         | Error handling policy, with options like 'NOOP', 'THROW', and 'RETRY'.   | THROW         |
| `connect.gcpstorage.http.max.retries`     | Number of HTTP request retries for resolvable errors.                  | 5             |
| `connect.gcpstorage.max.retries`         | Maximum number of write retries.                                       | 20            |
| `connect.gcpstorage.retry.interval`      | Time in milliseconds between retry attempts.                           | 60000         |
| `connect.gcpstorage.seek.max.files`       | Maximum index files allowed per topic/partition. Advisable not to raise this, as too many files indicate a problem with file deletion. | 5 |

## Getting Started

To get started with the Kafka sink connector for GCP Storage, you'll need to have a GCP account, create a GCP Storage Store, and obtain the necessary credentials. Then, configure the connector using the provided options to define the data flow from Kafka to GCP Storage.

Make sure to adjust the configuration according to your specific use case and requirements, and ensure that the GCP permissions are set up to allow data access and write operations to your GCP Storage storage.

In summary, this connector simplifies the process of streaming data from Kafka to GCP Storage, making it an essential tool for building real-time data pipelines and analytics solutions in the Azure ecosystem.

### CODE REVIEW

File: kafka-connect-gcp-storage/src/it/scala/io/lenses/streamreactor/connect/gcp/storage/utils/GCPProxyContainerTest.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    with UploadConfigKeys
+    with LazyLogging {
+
+  implicit val connectorTaskId: ConnectorTaskId = ConnectorTaskId("unit-tests", 1, 1)
+
+  override val container: PausableContainer = new TestContainersPausableContainer(GCPStorageContainer())
+
+  override val prefix: String = "connect.gcpstorage"
+
+  override def createStorageInterface(client: Storage): GCPStorageStorageInterface =
```
Comment (stheppi) [inline kafka-connect-gcp-storage/src/it/scala/io/lenses/streamreactor/connect/gcp/storage/utils/GCPProxyContainerTest.scala:40]: Is there an advantae to capture the exception go to an either and  fail if it's a Left? Since we don't handle the error let it bubble. ideally to avoid the Java approach we can return an Either or even a Try and let the caller handle it

File: kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/config/CommonConfigDef.scala [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+trait CommonConfigDef extends CompressionCodecConfigKeys with AuthModeSettingsConfigKeys {
+
+  def config: ConfigDef = {
+    val conf = new ConfigDef()
+      .define(
+        GCP_PROJECT_ID,
+        Type.STRING,
+        "",
+        Importance.HIGH,
+        "Azure Account Name",
```
Comment (stheppi) [inline kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/config/CommonConfigDef.scala:55]: We have a copy-paste issue here. It should say GCP. It applies to the other properties

File: kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/config/UploadSettings.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package io.lenses.streamreactor.connect.gcp.storage.config
+
+import com.datamountaineer.streamreactor.common.config.base.traits.BaseSettings
+import com.datamountaineer.streamreactor.common.config.base.traits.WithConnectorPrefix
+import org.apache.kafka.common.config.ConfigDef
+import org.apache.kafka.common.config.ConfigDef.Importance
+import org.apache.kafka.common.config.ConfigDef.Type
+trait UploadConfigKeys extends WithConnectorPrefix {
+  protected val AVOID_RESUMABLE_UPLOAD: String = s"$connectorPrefix.avoid.resumable.upload"
```
Comment (stheppi) [inline kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/config/UploadSettings.scala:24]: Let's document what the resumable enabled does and if there are any tradeoffs

File: kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/sink/GCPStorageWriterManagerCreator.scala [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import io.lenses.streamreactor.connect.cloud.common.sink.writer.WriterManager
+import io.lenses.streamreactor.connect.cloud.common.storage.StorageInterface
+import io.lenses.streamreactor.connect.gcp.storage.sink.config.GCPStorageSinkConfig
+import io.lenses.streamreactor.connect.gcp.storage.sink.config.SinkBucketOptions
+import io.lenses.streamreactor.connect.gcp.storage.storage.GCPStorageFileMetadata
+
+import java.io.File
+import scala.collection.immutable
+
+object GCPStorageWriterManagerCreator extends LazyLogging {
```
Comment (stheppi) [inline kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/sink/GCPStorageWriterManagerCreator.scala:43]: Is there much difference to the other writer creators? seems most code relies on bucket-for-topic mapping and that's the same across all cloud storage configs. Maybe we can avoid duplicating code even more

File: kafka-connect-gcp-storage/src/it/scala/io/lenses/streamreactor/connect/gcp/storage/utils/GCPProxyContainerTest.scala (hunk 1 above)
Comment (davidsloan) [reply kafka-connect-gcp-storage/src/it/scala/io/lenses/streamreactor/connect/gcp/storage/utils/GCPProxyContainerTest.scala:40]: No advantage.  I like the second approach you suggest.  The code as it is is ultimately a mixture of styles so not easy to get consistency without a lot of further work.  I've made some tweaks to improve the situation a bit.

File: kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/sink/GCPStorageWriterManagerCreator.scala (hunk 4 above)
Comment (davidsloan) [reply kafka-connect-gcp-storage/src/main/scala/io/lenses/streamreactor/connect/gcp/storage/sink/GCPStorageWriterManagerCreator.scala]: 💯  - these 3 have become 1 and it was much less painful than expected.  Thanks!

**Review by stheppi — APPROVED** [review]
